File: sdk/typescript/src/items.ts
/**
 * Item types representing content within a session turn.
 *
 * Each item corresponds to a specific `SdkMessage` variant from the Rust CLI.
 */

// ---------------------------------------------------------------------------
// Content blocks (mirrors Rust ContentBlock enum)
// ---------------------------------------------------------------------------

export type TextBlock = { type: "text"; text: string };
export type ToolUseBlock = {
  type: "tool_use";
  id: string;
  name: string;
  input: unknown;
};
export type ToolResultBlock = {
  type: "tool_result";
  tool_use_id: string;
  content: unknown;
  is_error: boolean;
};
export type ThinkingBlock = {
  type: "thinking";
  thinking: string;
  signature?: string;
};
export type RedactedThinkingBlock = { type: "redacted_thinking"; data: string };
export type ImageBlock = { type: "image"; source: ImageSource };

export type ContentBlock =
  | TextBlock
  | ToolUseBlock
  | ToolResultBlock
  | ThinkingBlock
  | RedactedThinkingBlock
  | ImageBlock;

export type ImageSource = {
  type: string;
  media_type: string;
  data: string;
};

// ---------------------------------------------------------------------------
// Message-level usage
// ---------------------------------------------------------------------------

export type MessageUsage = {
  input_tokens: number;
  output_tokens: number;
  cache_read_input_tokens: number;
  cache_creation_input_tokens: number;
};

// ---------------------------------------------------------------------------
// Item types
// ---------------------------------------------------------------------------

/** Agent message — from `SdkMessage::Assistant`. */
export type AgentMessageItem = {
  id: string;
  type: "agent_message";
  text: string;
  content_blocks: ContentBlock[];
  usage?: MessageUsage;
  stop_reason?: string;
  cost_usd: number;
};

/** Tool-use summary — from `SdkMessage::ToolUseSummary`. */
export type ToolUseSummaryItem = {
  id: string;
  type: "tool_use_summary";
  summary: string;
  preceding_tool_use_ids: string[];
};

/** Compact boundary — from `SdkMessage::CompactBoundary`. */
export type CompactBoundaryItem = {
  id: string;
  type: "compact_boundary";
  pre_compact_token_count?: number;
  post_compact_token_count?: number;
};

/** User message replay — from `SdkMessage::UserReplay`. */
export type UserReplayItem = {
  id: string;
  type: "user_replay";
  content: string;
  is_replay: boolean;
  is_synthetic: boolean;
};

/** Error item. */
export type ErrorItem = {
  id: string;
  type: "error";
  message: string;
};

// ---------------------------------------------------------------------------
// Union
// ---------------------------------------------------------------------------

export type SessionItem =
  | AgentMessageItem
  | ToolUseSummaryItem
  | CompactBoundaryItem
  | UserReplayItem
  | ErrorItem;

File: sdk/python/src/claude_code_rs/items.py
"""Item types representing content within a session turn.

Mirrors TypeScript SDK's items.ts. Each item corresponds to a specific
``SdkMessage`` variant from the Rust CLI.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Union

# ---------------------------------------------------------------------------
# Content blocks (mirrors Rust ContentBlock enum)
# ---------------------------------------------------------------------------


@dataclass
class TextBlock:
    type: str  # "text"
    text: str


@dataclass
class ToolUseBlock:
    type: str  # "tool_use"
    id: str
    name: str
    input: Any


@dataclass
class ToolResultBlock:
    type: str  # "tool_result"
    tool_use_id: str
    content: Any
    is_error: bool = False


@dataclass
class ThinkingBlock:
    type: str  # "thinking"
    thinking: str
    signature: str | None = None


@dataclass
class RedactedThinkingBlock:
    type: str  # "redacted_thinking"
    data: str


@dataclass
class ImageSource:
    type: str
    media_type: str
    data: str


@dataclass
class ImageBlock:
    type: str  # "image"
    source: ImageSource


ContentBlock = Union[
    TextBlock,
    ToolUseBlock,
    ToolResultBlock,
    ThinkingBlock,
    RedactedThinkingBlock,
    ImageBlock,
]

# ---------------------------------------------------------------------------
# Message-level usage
# ---------------------------------------------------------------------------


@dataclass
class MessageUsage:
    input_tokens: int = 0
    output_tokens: int = 0
    cache_read_input_tokens: int = 0
    cache_creation_input_tokens: int = 0


# ---------------------------------------------------------------------------
# Item types
# ---------------------------------------------------------------------------


@dataclass
class AgentMessageItem:
    """Agent message — from ``SdkMessage::Assistant``."""

    id: str
    type: str = "agent_message"
    text: str = ""
    content_blocks: list[ContentBlock] = field(default_factory=list)
    usage: MessageUsage | None = None
    stop_reason: str | None = None
    cost_usd: float = 0.0


@dataclass
class ToolUseSummaryItem:
    """Tool-use summary — from ``SdkMessage::ToolUseSummary``."""

    id: str
    type: str = "tool_use_summary"
    summary: str = ""
    preceding_tool_use_ids: list[str] = field(default_factory=list)


@dataclass
class CompactBoundaryItem:
    """Compact boundary — from ``SdkMessage::CompactBoundary``."""

    id: str
    type: str = "compact_boundary"
    pre_compact_token_count: int | None = None
    post_compact_token_count: int | None = None


@dataclass
class UserReplayItem:
    """User message replay — from ``SdkMessage::UserReplay``."""

    id: str
    type: str = "user_replay"
    content: str = ""
    is_replay: bool = False
    is_synthetic: bool = False


@dataclass
class ErrorItem:
    """Error item."""

    id: str
    type: str = "error"
    message: str = ""


SessionItem = Union[
    AgentMessageItem,
    ToolUseSummaryItem,
    CompactBoundaryItem,
    UserReplayItem,
    ErrorItem,
]
